northstar/src/api/codec.rs

```diff
@@ -0,0 +1,144 @@
+                Err(e) => Err(io::Error::new(ErrorKind::Other, e)),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl Encoder<model::Message> for Codec {
+    type Error = io::Error;
+
+    fn encode(
+        &mut self,
+        item: model::Message,
+        dst: &mut bytes::BytesMut,
+    ) -> Result<(), Self::Error> {
+        dst.extend_from_slice(serde_json::to_string(&item)?.as_bytes());
+        dst.reserve(1);
+        dst.put_u8(b'\n');
+        Ok(())
+    }
+}
+
+impl<T: Unpin + AsyncRead + AsyncWrite> AsyncWrite for Framed<T> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let t: &mut T = self.framed.get_mut();
+        AsyncWrite::poll_write(Pin::new(t), cx, buf)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        let t: &mut T = self.framed.get_mut();
+        AsyncWrite::poll_flush(Pin::new(t), cx)
+    }
+
+    fn poll_shutdown(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), io::Error>> {
+        let t: &mut T = self.framed.get_mut();
+        AsyncWrite::poll_shutdown(Pin::new(t), cx)
+    }
+}
+
+impl<T: Unpin + AsyncRead + AsyncWrite> AsyncRead for Framed<T> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut io::ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        if self.framed.read_buffer().is_empty() {
+            let t: &mut T = self.framed.get_mut();
+            AsyncRead::poll_read(Pin::new(t), cx, buf)
+        } else {
+            let n = min(buf.remaining(), self.framed.read_buffer().len());
+            buf.put_slice(&self.framed.read_buffer_mut().split_to(n));
+            Poll::Ready(Ok(()))
+        }
+    }
+}
+
+impl<T: Unpin + AsyncWrite + AsyncRead> Stream for Framed<T> {
+    type Item = Result<model::Message, io::Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let framed = Pin::new(&mut self.framed);
+        framed.poll_next(cx)
+    }
+}
+
+impl<T: Unpin + AsyncRead + AsyncWrite> futures::sink::Sink<model::Message> for Framed<T> {
+    type Error = io::Error;
+
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Pin::new(&mut self.framed).poll_ready(cx)
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: model::Message) -> Result<(), Self::Error> {
+        Pin::new(&mut self.framed).start_send(item)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Pin::new(&mut self.framed).poll_flush(cx)
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Pin::new(&mut self.framed).poll_close(cx)
+    }
+}
```

Comment:
The `AsyncWrite`, `AsyncRead`, `Sink` and `Stream` implementation looks a lot like boilerplate code. My understanding of the codec package was that you can use `FramedRead` and `FramedWrite` to achieve this.

Reply:
`FramedWrite` and `AsyncWrite` work different. If we want to use `FramedWrite` (mean `Sink`) for sending the install stream we need to introduce another enum variant for `Framed::Item` and chunk manually. This is safe but probably slower than this direct access to the wrapped `AsyncWrite`. The intention to newtype `Framed` is to use it as `Stream`, `Sink` *and* `AsyncWrite`.